### PR TITLE
Add ImpactX

### PR DESCRIPTION
**Pull request by ax3l**
Add ImpactX: a modern, s-based beam dynamics code for particle accelerator research and design.

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [ ] Package does not vendor other packages. [AMReX](https://github.com/AMReX-Codes/amrex) & [pyAMReX](https://github.com/AMReX-Codes/pyamrex) are pulled on the fly, under the same license. Many binary variants make separating them very hard for now.
- [x] If static libraries are linked in, the license of the static library is packaged.
- [ ] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

File: recipes/impactx/bld.bat [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+:: build
+cmake --build build --config RelWithDebInfo --parallel 2
+if errorlevel 1 exit 1
+cmake --build build --config RelWithDebInfo --parallel 2 --target pyamrex_pip_wheel
+if errorlevel 1 exit 1
+cmake --build build --config RelWithDebInfo --parallel 2 --target pip_wheel
+if errorlevel 1 exit 1
+
+:: test
+ctest --test-dir build --output-on-failure -E AMReX
```
Comment (ax3l) [inline recipes/impactx/bld.bat:35]:
```suggestion
ctest --test-dir build --build-config RelWithDebInfo --output-on-failure -E AMReX
```

**Comment by ax3l** [comment]
@isuruf since you merged WarpX in #13273, can I bother you again? :)

**Comment by ax3l** [comment]
@conda-forge/help-c-cpp, ready for review & merge :)

File: recipes/impactx/build.sh [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+# IPO/LTO does only work with certain toolchains
+WarpX_IPO=ON
+if [[ ${target_platform} =~ osx.* ]]; then
+    WarpX_IPO=OFF
+fi
+
+# configure
+cmake \
+    -S ${SRC_DIR} -B build                \
```
Comment (isuruf) [inline recipes/impactx/build.sh:22]:
```suggestion
    -S ${SRC_DIR} -B build                \
    ${CMAKE_ARGS} \
```

File: recipes/impactx/build.sh [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+# Darwin modern C++
+#   https://conda-forge.org/docs/maintainer/knowledge_base.html#newer-c-features-with-old-sdk
+if [[ ${target_platform} =~ osx.* ]]; then
+    export CXXFLAGS="${CXXFLAGS} -D_LIBCPP_DISABLE_AVAILABILITY"
+fi
+
+# IPO/LTO does only work with certain toolchains
+WarpX_IPO=ON
+if [[ ${target_platform} =~ osx.* ]]; then
+    WarpX_IPO=OFF
```
Comment (isuruf) [inline recipes/impactx/build.sh:17]: IPO doesn't work on macOS?

File: recipes/impactx/build.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+cmake \
+    -S ${SRC_DIR} -B build                \
+    -DCMAKE_BUILD_TYPE=RelWithDebInfo     \
+    -DCMAKE_VERBOSE_MAKEFILE=ON           \
+    -DCMAKE_INSTALL_LIBDIR=lib            \
+    -DCMAKE_INSTALL_PREFIX=${PREFIX}      \
+    -DImpactX_COMPUTE=NOACC               \
+    -DImpactX_IPO=${ImpactX_IPO}          \
+    -DImpactX_amrex_branch=22.08          \
+    -DImpactX_pyamrex_branch=c11acfaf08162fe42c5ee07c086d23f2874fa779 \
```
Comment (isuruf) [inline recipes/impactx/build.sh:30]: Is the license of amrex and pyamrex included in `license_file`?

File: recipes/impactx/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    - pip
+    - libgomp       # [linux]
+    - llvm-openmp   # [osx or win]
+    # ctest
+    - matplotlib-base  # [build_platform == target_platform]
+    - numpy            # [build_platform == target_platform]
+    - pandas           # [build_platform == target_platform]
+    - pytest           # [build_platform == target_platform]
+    - python           # [build_platform == target_platform]
+    - scipy            # [build_platform == target_platform]
```
Comment (isuruf) [inline recipes/impactx/meta.yaml:46]: Why are these in `build` and not `host`?

File: recipes/impactx/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    - {{ compiler('cxx') }}
+    - ccache        # [unix]
+    - clang         # [win]
+    - cmake >=3.20.0
+    - lld           # [win]
+    # TODO: add libxml2 dep in lld-feedstock
+    - libxml2       # [win]
+    - make          # [unix]
+    - ninja         # [win]
+    - pip
```
Comment (isuruf) [inline recipes/impactx/meta.yaml:37]: This should probably be in `host` as well.

File: recipes/impactx/build.sh (hunk 4 above)
Comment (ax3l) [reply recipes/impactx/build.sh:30]: They use the same license.

File: recipes/impactx/build.sh (hunk 3 above)
Comment (ax3l) [reply recipes/impactx/build.sh:17]: Not the last time I tried with WarpX

File: recipes/impactx/meta.yaml (hunk 5 above)
Comment (ax3l) [reply recipes/impactx/meta.yaml:46]: I tried in `host` and it did not find some of the packages at `ctest` time

File: recipes/impactx/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+  doc_url: https://impactx.readthedocs.io
+  dev_url: https://github.com/ECP-WarpX/impactx
+
+extra:
+  recipe-maintainers:
+    - ax3l
+    - cemitch99
+    - RemiLehe
+    - RTSandberg
```
Comment (ax3l) [inline recipes/impactx/meta.yaml:98]: @cemitch99 @RemiLehe @RTSandberg can you please confirm here with a comment that you agree to be conda-forge co-maintainers for ImpactX? :)

File: recipes/impactx/build.sh (hunk 2 above)
Comment (ax3l) [reply recipes/impactx/build.sh:22]: added :+1:

File: recipes/impactx/meta.yaml (hunk 6 above)
Comment (ax3l) [reply recipes/impactx/meta.yaml]: moved, let's see if we can get it all working :)

File: recipes/impactx/meta.yaml (hunk 5 above)
Comment (ax3l) [reply recipes/impactx/meta.yaml:51]: moved, let's see if we can get it all working :)

File: recipes/impactx/meta.yaml (hunk 7 above)
Comment (RTSandberg) [reply recipes/impactx/meta.yaml:97]: ok with me!

Comment (n01r) [reply recipes/impactx/meta.yaml:97]: I'm also be in :)

Comment (cemitch99) [reply recipes/impactx/meta.yaml:97]: Agreed.

Comment (RemiLehe) [reply recipes/impactx/meta.yaml:97]: OK

**Comment by ax3l** [comment]
macOS:
```
      Start 28: pytest.ImpactX
28/28 Test #28: pytest.ImpactX ...................***Failed    0.07 sec
/Applications/Xcode_13.2.1.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.8/bin/python3.8: No module named pytest
```

wild.

File: recipes/impactx/build.sh [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    -DImpactX_LIB=ON      \
+    -DImpactX_MPI=OFF     \
+    -DImpactX_PYTHON=ON   \
+    -DPython3_FIND_STRATEGY=LOCATION
```
Comment (ax3l) [inline recipes/impactx/build.sh:35]:
```suggestion
    -DPython3_EXECUTABLE=${PYTHON}
```

File: recipes/impactx/bld.bat [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    -DCMAKE_CXX_COMPILER=clang-cl         ^
+    -DCMAKE_LINKER=lld-link               ^
+    -DCMAKE_NM=llvm-nm                    ^
+    -DCMAKE_VERBOSE_MAKEFILE=ON           ^
+    -DImpactX_amrex_branch=22.08          ^
+    -DImpactX_pyamrex_branch=c11acfaf08162fe42c5ee07c086d23f2874fa779 ^
+    -DImpactX_COMPUTE=NOACC ^
+    -DImpactX_LIB=ON        ^
+    -DImpactX_MPI=OFF       ^
+    -DImpactX_PYTHON=ON
```
Comment (ax3l) [inline recipes/impactx/bld.bat:24]:
```suggestion
    -DImpactX_MPI=OFF       ^
    -DImpactX_PYTHON=ON     ^
    -DPython_EXECUTABLE=%PYTHON%
```

**Comment by ax3l** [comment]
@isuruf wuhu, ready to be merged :)

**Comment by ax3l** [comment]
@isuruf ready to be merged :)

File: recipes/impactx/build.sh (hunk 4 above)
Comment (isuruf) [reply recipes/impactx/build.sh:31]: Do they have the same copyright holder as well?

Comment (ax3l) [reply recipes/impactx/build.sh:31]: Yes, in both cases LBNL/UC

**Pull request merged**